Extended Calendar

xXFiEsTaDeAmOnXx opened the pull request:
## Describe your changes


As mentioned in [Issue #1853](https://github.com/end-4/dots-hyprland/issues/1853), todo items should have a specific date when they expire.  
I implemented a method to set a **deadline** for each todo item, along with a visualization.  
In the calendar view, clicking on a day now displays the tasks assigned to that date.  


 



| Select Deadline | Todo List | Calendar (no hover) | Calendar (with hover) |
|-----------------|-----------|---------------------|-----------------------|
|<img width="450" height="1436" alt="image" src="https://github.com/user-attachments/assets/8a5779c4-d6b9-4826-9cc3-4d78849dec64" />|<img width="452" height="1440" alt="image" src="https://github.com/user-attachments/assets/487cc677-7b62-428f-b22e-43dbbadffde8" />|<img width="454" height="1433" alt="image" src="https://github.com/user-attachments/assets/b858f2f3-65c9-4d85-aec2-e9599a7427dd" />|<img width="461" height="1440" alt="image" src="https://github.com/user-attachments/assets/b331f219-c31b-4fd7-bc58-ac7e4ef031ff" />|

### Future work
Synchronization with Google Calendar and Google Tasks could be added in a future PR, depending on demand.  

---

### Is it ready? / Feedback needed
Looking for feedback on whether the feature works as expected and if the UX is clear.

xXFiEsTaDeAmOnXx commented:
Alright I will look into this.

xXFiEsTaDeAmOnXx commented:
- I added Khal integration.  
- You can now add and delete events within khal directly from the UI.  
- You will see all khal events for the next 3 month in your calendar.
- The todo state will **not** be saved in Khal, so it is temporary.  
- Since I switched to the LazyLoader, the placement of the PopUp can be a bit off when multiple Events appear on the same day. I haven’t found a fix for this yet.
- If Khal is installed, it will be used instead of the local JSON file.  
- Deleting events is somewhat hacky, as there is currently no way to remove them without using the interactive shell.  Therefore I have to use sqlite3 directly.
  See issue: [khal #603](https://github.com/pimutils/khal/issues/603)

xXFiEsTaDeAmOnXx commented:
Yeah, you  currently have to scroll to select a deadline. I know this is not the best UX design, but I did not have the time to implement a Datepicker yet as there is none.

To the population I cannot say anything.  I currently I am only selecting the events within 3 month in the future.

You can use this command to try it  in shell to see if it returns an event.

```bash
khal list --json title --json start-date 09/09/2025 09/12/2025
```

If it does not return an event, the calendar will also not show an event.

xXFiEsTaDeAmOnXx commented:
I will look into it when I got time.

xXFiEsTaDeAmOnXx commented:
Yeah, I have this issue too.

I’m not sure where it comes from. Is this happening in the main branch as well?

It might be related to the second window being overlaid on top, but I don’t think that’s the actual issue. It shouldn’t even be there, since it should be lazy-loaded.

xXFiEsTaDeAmOnXx commented:
Alright, I added it into the cheetsheet.

Preview:

<img width="1270" height="887" alt="image" src="https://github.com/user-attachments/assets/c2c4963e-925b-4598-8a76-f7c27664e642" />

I still need to do some testing and refactoring.

xXFiEsTaDeAmOnXx commented:
Okay, I did not notice it as I was using 24 hour clock . Is the issue now fixed?

xXFiEsTaDeAmOnXx commented:
Yes google calendar works for me.

### Creating the config for vdirsycner

I used the vdirsycner config provided in this [blog post](https://aymanbagabas.com/blog/2018/04/08/sync-google-calendar.html):


```
[general]
status_path = "~/.vdirsyncer/status/"

[pair calendar]
a = "google_calendar"
b = "local_calendar"
collections = ["from a", "from b"]
metadata = ["color"]

[storage local_calendar]
type = "singlefile"
path = "~/.calendars/%s.ics"

[storage google_calendar]
type = "google_calendar"
token_file = "~/.vdirsyncer/google_token"
client_id = "CLIENT_ID"
client_secret = "CLIENT_SECRET"
```

### Obtaining `CLIENT_ID` and `CLIENT_SECRET`**


I did the following described in the [vdirsycner docs](https://vdirsyncer.pimutils.org/en/stable/config.html):

1. Go to the [Google API Manager](https://console.developers.google.com/)

2. Create a new project under any name.

3. Within that project, enable the “CalDAV” . There should be a search box where you can just enter those terms.

4. In the sidebar, select “Credentials”, then “Create Credentials” and create a new “OAuth Client ID”.

4. You’ll be prompted to create a OAuth consent screen first. Fill out that form however you like.

5. After setting up the consent screen, finish creating the new “OAuth Client ID’. The correct application type is “Desktop application”.

6. Finally you should have a Client ID and a Client secret. Provide these in your storage config.

### Authorizing vdirsyncer

After that you can try giving vdirsyncer acess to your google calender by using the following command.
```
 vdirsyncer discover
```
This command should open your browser. You can give the vdirsyncer access to your account via the browser

**Note:** You should also  make sure that your previously created project within Google API Manager is in production and not in testing stage under the sidebar audience tab in your sidebar. Otherwise this step fails and google complains your are not invited to test the application.

xXFiEsTaDeAmOnXx commented:
Did you set your product  to 'in production' within audience?

<img width="960" height="202" alt="image" src="https://github.com/user-attachments/assets/e4a959ff-9e74-4f46-918a-1d72df0303ab" />

This was the thing that fixed the error for me.

webcubed commented:
so i set up vdirsyncer just fine but khal refuses to pick up on the vdirsyncer calendars

unlimitedsola committed:
fix overview scaling issue

OfekiAlm commented:
> Sure, you can try fix inside the `CheatsheetTimetable.qml`, if you are comfortable with qml.

Great! I PRed into your branch on your forked repo
since I don't have permission to push commits here  😢 
review it and lets think how we can make it better :)

xXFiEsTaDeAmOnXx committed:
Removed todo deadline and calendar integration

xXFiEsTaDeAmOnXx committed:
Redisigned allday events

OfekiAlm commented:
Really nice work dude! How would it look with multiple all-day events?

xXFiEsTaDeAmOnXx committed:
Fixed multiple day hovering

xXFiEsTaDeAmOnXx committed:
Used primary color for calendar indicator

xXFiEsTaDeAmOnXx committed:
Indroduced first day of week config

xXFiEsTaDeAmOnXx committed:
Bug fixes

xXFiEsTaDeAmOnXx committed:
Bug fixes

xXFiEsTaDeAmOnXx committed:
Fixed screen when khal is not installed

xXFiEsTaDeAmOnXx committed:
Switched firstDayOfWeek  back to monday

xXFiEsTaDeAmOnXx committed:
Merge branch 'main' into calendar

clsty committed:
Updata uv/README.md

Cleboost committed:
Add screen recording button and config toggle

Cleboost committed:
fix: Invoke record script via bash -c with ~ expansion

BeshoyEhab committed:
Enhance the update script

BeshoyEhab committed:
Fix the local used outside a function issue

BeshoyEhab committed:
Update the Warning message

BeshoyEhab committed:
Add test_update sctipt

BeshoyEhab committed:
Fix some security issues and remove non-used function

BeshoyEhab committed:
Update the test script

BeshoyEhab committed:
Fix some issues in the test_update script

Alvin-HZ committed:
Fix escape keybind notification for virtual machines

0blivi0nis committed:
✨ feat: sound alerts for battery and pomodoro

clsty committed:
Fix backup function typo

clsty committed:
Fix diagnose

BeshoyEhab committed:
Integrate update script into install.sh

clsty committed:
Try to fix backup

BeshoyEhab committed:
Update the update and options and install to correctly pass the args to update script

clsty committed:
Use local var for args_includes

BeshoyEhab committed:
Update install script to pass the update args

BeshoyEhab committed:
Enhance the logic for pass the args to update script

BeshoyEhab committed:
Maybe something fixed?

BeshoyEhab committed:
Test

BeshoyEhab committed:
Test

BeshoyEhab committed:
Test

BeshoyEhab committed:
Test

BeshoyEhab committed:
Test

BeshoyEhab committed:
Return the old update script

BeshoyEhab committed:
Fix: Prevent script exit in while-read loop

BeshoyEhab committed:
Fix: Suppress /dev/tty error in safe_read

BeshoyEhab committed:
Remove a useless condition

end-4 committed:
sidebar: add scroll to bottom button

end-4 committed:
fix settings app cant change time format

clsty committed:
Better integration; introduce subcommand

clsty committed:
Rename scripts, move into sdata/step

clsty committed:
Improve/Fix showhelp

end-4 committed:
fix gemini clock theming (forgor mkdir)

end-4 committed:
sidebar: android-style quick toggles & sliders (#2217)

Co-Authored-By: Vague Syntax <173799252+vaguesyntax@users.noreply.github.com>

end-4 committed:
sidebar: quick toggles: fix iilegal default

end-4 committed:
sidebar: make status text of some toggles make more sense

BeshoyEhab committed:
Test the changes that have been made

end-4 committed:
sidebar: quick toggles: add missing tooltips

end-4 committed:
readme: update "some widgets" screenshot

end-4 committed:
sidebar: quick toggles: fix weird anim on startup

BeshoyEhab committed:
Update the update, update tester, functions scripts

BeshoyEhab committed:
Test

BeshoyEhab committed:
Fixed the test and update script

BeshoyEhab committed:
Maybe fix the smoke-test

BeshoyEhab committed:
Remove the smoke-test

BeshoyEhab committed:
Fix a UI issue

clsty committed:
Update showfun()

clsty committed:
Fixes for install scripts

clsty committed:
Using ./setup install instead of ./install.sh

clsty committed:
Fix subcmd exp-uninstall --help

clsty committed:
Update message

clsty committed:
Update comment

clsty committed:
Update comment

clsty committed:
Update diagnose

clsty committed:
Deprecate install.sh in favor of setup

Still decide to deprecate install.sh cuz TAB completion can not be done
without intrusive and automatic method, so just use ./setup instead.

clsty committed:
Update showhelp_global()

clsty committed:
Update showhelp

clsty committed:
Pin commit for quickshell-git (#2238)

clsty committed:
Update comment in PKGBUILD

clsty committed:
Ask once instead of 3 times for backup

clsty committed:
Add -f for makepkg

end-4 committed:
resourcePopup: fix unqualified access

end-4 committed:
roundcorner: use switch fallthrough for fewer dupe returns

end-4 committed:
settings: add api key note for ai translation

end-4 committed:
utilbuttons: record button: allow region selection

end-4 committed:
bar: record util button off by default

end-4 committed:
settings: make bar screen record util button toggle name shorter

end-4 committed:
qs: use more neutral default pallete

end-4 committed:
fix sound plays every time low and charge state changes

instead of just when it goes to low

end-4 committed:
right sidebar: move audio controls to dialogs

end-4 committed:
update default quick toggles

end-4 committed:
make notifs dismissable in both directions

end-4 committed:
make notif items also draggable to left

Yunexiz committed:
Show passive items in the system tray

end-4 committed:
keybinds: make office keybind match windows, allow record when locked

end-4 committed:
make screen snip built into the shell so it's faster (#2113)

end-4 committed:
hyprland: update keybinds+rules for previous commit

Yunexiz committed:
add config option for smartTray

Yunexiz committed:
typo

end-4 committed:
circle to search

end-4 committed:
region selector: fix target regions

end-4 committed:
make google lens tool use normal region selection by default

end-4 committed:
right sidebar: containerize top row elements

end-4 committed:
sidebar: quick toggles: smoother size change

end-4 committed:
sidebar: quick toggles: larger icon size when 1 cell wide

jwihardi committed:
added live quickshell ebuild with specific commit

jwihardi committed:
fixed package name

jwihardi committed:
Delete sdist/gentoo/illogical-impulse-quickshell-git/quickshell-9999.ebuild

jwihardi committed:
quickshell, widgets, install, useflag (breakpad), keywords - updated

jwihardi committed:
gui-misc -> app-misc

clsty committed:
Add --skip-quickshell for install-files.sh

clsty committed:
Update package-installers.sh

clsty committed:
Add auto-close-issue workflow (WIP)

clsty committed:
Try to fix auto-close-issue.yml

clsty committed:
Try to fix auto-close-issue.yml

clsty committed:
Try to fix auto-close-issue.yml

clsty committed:
Try to fix auto-close-issue.yml

clsty committed:
Try to fix auto-close-issue.yml

clsty committed:
Improve auto-close-issue.yml

clsty committed:
Finish adding auto-close-issue.yml

clsty committed:
Update message

end-4 committed:
allow corner hover open for horizontal bar

end-4 committed:
make left sidebar padding consistent with right sidebar

end-4 committed:
fix screenshot button

end-4 committed:
translation?service?:safer?access???

end-4 committed:
fix kill dialog not writing "always" properly (#2232)

end-4 committed:
hide code: entries from cheatsheet for zoom keybinds

end-4 committed:
sidebar: fix weird anim on toggle buttons

end-4 committed:
fix "no active player" popup positioning for vertical bar

janusfil committed:
refactor import to be compatible with new version of quickshell

janusfil committed:
fix bar module reference

end-4 committed:
remove redundant imports

end-4 committed:
update qs version

end-4 committed:
refractor icon toolbar buttons

end-4 committed:
make sidebar classic toggle button icons have size consistent with toolbar

end-4 committed:
revamp region selector

end-4 committed:
adjust screenshot regions opacity

clsty committed:
Update quickshell version (gentoo)

end-4 committed:
make region selector more readable in light theme

end-4 committed:
region selector: animate icon on mode change

end-4 committed:
systeminfo: add gentoo

end-4 committed:
brightness service: more edging

end-4 committed:
sidebar: unfuck non-default quick toggles configurations

end-4 committed:
styledslider: make marks with any range

end-4 committed:
sidebar: night light dialog

clsty committed:
Not overwrite other things under quickshell

clsty committed:
Add subcmds; Not skip backup unless input y/n

end-4 committed:
fix deprecated Connections syntax

end-4 committed:
add anti flashbang

could be improve with smooth fading

end-4 committed:
region selector: add option to show/hide coord aim lines

end-4 committed:
region selector: lower aim lines opacity

end-4 committed:
anti flashbang: clean up when done

Yunexiz committed:
change name to "filterPassive"

end-4 committed:
anti flashbang: pass image directly to magick thru stdio, more triggers

end-4 committed:
brightness: animate change for laptop screens, prevent pitch black w/ anti flashbang

end-4 committed:
make night light color temp slider more intuitive

clsty committed:
Remove .gitattributes

end-4 committed:
region selector: cleanup after image search

end-4 committed:
notif groups: expand on right press instead of release

end-4 committed:
work safety: disabled by default, add toggle to settings app

end-4 committed:
workspace_action.sh: format

end-4 committed:
add fallback script for google lens keybind

end-4 committed:
brightness: adjust antiflashbang delays, indicate nightlight on osd

end-4 committed:
settings: add ws number map presets

end-4 committed:
settings: add roman ws number preset

end-4 committed:
use quickshell region selector for ocr

end-4 committed:
region selector: make circle selection not amogus large

end-4 committed:
region selector: fix circle dragging being forced into a target region

end-4 committed:
ocr: use all available langs

end-4 committed:
region selector: fix dismiss button

end-4 committed:
right sidebar: fix unused quick toggles being movable

clsty committed:
Update sdist/nix/README.md

end-4 committed:
sidebar: use friendlier name for idle inhibitor toggle

end-4 committed:
don't create homework folder before it's necessary (#2264)

end-4 committed:
fix calendar occasionally get placed at funny position

end-4 committed:
region selector: remove auto fullscreen ocr, make close btn fab

ocr-ing the whole screen slow and messy. it's better to just go with the regular flow of selecting a region

end-4 committed:
make notif status row less weird

end-4 committed:
make notif action buttons larger

clsty committed:
(WIP) Bring back old update.sh (#2284)

clsty committed:
Update exp-update-old.sh

clsty committed:
Bring back old update.sh (#2284)

clsty committed:
Fix path in exp-update-old.sh

clsty committed:
Fix -s for exp-update

end-4 committed:
sidebar: make notif delete icon more intuitive

end-4 committed:
adjust fab icon size and region selector fab color

xXFiEsTaDeAmOnXx committed:
Merge branch 'main' into calendar

xXFiEsTaDeAmOnXx committed:
Fixed import

xXFiEsTaDeAmOnXx committed:
Merge branch 'main' into calendar

xXFiEsTaDeAmOnXx committed:
Merge branch 'main' into calendar

xXFiEsTaDeAmOnXx committed:
Merge branch 'main' into calendar

xXFiEsTaDeAmOnXx committed:
Merge branch 'main' into calendar

xXFiEsTaDeAmOnXx committed:
Merge branch 'main' into calendar

xXFiEsTaDeAmOnXx committed:
Fixed overview offset

xXFiEsTaDeAmOnXx committed:
Fix calendar offset

xXFiEsTaDeAmOnXx committed:
Merge branch 'main' into calendar

Reluckycf committed:
change hyprsunset default color temp to 6000

montanohanz committed:
feat: update emoji list to emoji 17.0

Linuxius-Thomas committed:
Add files via upload

Linuxius-Thomas committed:
Create main.lua

Linuxius-Thomas committed:
Add files via upload

Linuxius-Thomas committed:
Create env.lua

Linuxius-Thomas committed:
Create execs.lua

Linuxius-Thomas committed:
Create general.lua

Linuxius-Thomas committed:
Create keybinds.lua

Linuxius-Thomas committed:
Create rules.lua

Linuxius-Thomas committed:
Create variables.lua

Linuxius-Thomas committed:
Add files via upload

Linuxius-Thomas committed:
Add files via upload

Linuxius-Thomas committed:
Add files via upload

Linuxius-Thomas committed:
Add files via upload

Linuxius-Thomas committed:
Add files via upload

Linuxius-Thomas committed:
Add files via upload

fb24m committed:
fix kitty config

fb24m committed:
rewrite script to use RPM built using GitHub action

fb24m committed:
replace tested note

xXFiEsTaDeAmOnXx committed:
Merge branch 'main' into calendar

electricsteve committed:
feat: notification on specific monitor

Linuxius-Thomas committed:
Update general.lua

end-4 committed:
fix stuff

end-4 committed:
use globalshortcut for 4finger up touchpad gesture

end-4 committed:
qs: fix hyprland dispatchers

end-4 committed:
remove redundant wallpaper switching stuff

end-4 committed:
qs: fix hyprland overrides (game mode, anti flashbang)

end-4 committed:
fix screen locking

end-4 committed:
qs: fix some manual hyprland dispatches

end-4 committed:
fix keybind cheatsheet

end-4 committed:
hyprland: exclude monitors and workspaces conf

end-4 committed:
previous commit but in the "legacy" script

end-4 committed:
install script: rename hyprland.conf to not use it

end-4 committed:
install script: add check for hyprland.conf rename

end-4 committed:
add example keybind for editing user keybinds

end-4 committed:
fix hyprland theming

end-4 committed:
readme: add hyprland 0.55 warning

end-4 committed:
qs: temp fix super key always thought to be held down

end-4 committed:
qs: fix super key hold state properly

Rasmion committed:
fix(keybinds): correct fullscreen and maximize bind syntax

end-4 committed:
readme: move warning content to wiki

Rasmion committed:
fix: specify toggle action explicitly

imitoy committed:
Fix HOME env in lua

GregorVal committed:
Do not override XDG_DATA_DIRS (fixes #2583)

end-4 committed:
qs: rename config option notifications.monitor to notifications.forceMonitor

end-4 committed:
hl: keybinds: fix workspace groups, categorize descriptions

end-4 committed:
hl config: remove unused bash scripts

end-4 committed:
qs: cheatsheet: display categories nicely

end-4 committed:
ctrl+super+shift+d for dark/light toggle (like powertoys)

end-4 committed:
hl: create custom config files automatically

end-4 committed:
hl: fix #3327 send to scratchpad desc at wrong scope

end-4 committed:
hl: remove that whatever jetbrains windowrule (fixes #3324)

end-4 committed:
hl: make custom stuff optional

end-4 committed:
qs: fix wrong capitalization in cheatsheet

end-4 committed:
qs: remove no longer necessary unlock refocus hack

end-4 committed:
remove redundant scripts

VietNguyenVN committed:
Update keybinds.lua

end-4 committed:
add back some repetitive keybinds to cheatsheet

zzallirog committed:
fix(hyprsunset): remove vestigial Hyprland.dispatch broken under .lua schema

GregorVal committed:
Fix: XDG_DATA_DIRS now expands correctly

Previously the environmental variable would become literally
*:$XDG_DATA_DIRS
Now the variable is expanded correctly
